File: app/src/main/java/podtest/jiki/se/podtest/start/PagedEpisodesAdapter.java
package podtest.jiki.se.podtest.start;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import podtest.jiki.se.podtest.LazyLoadedEpisodeList;
import podtest.jiki.se.podtest.R;


public class PagedEpisodesAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    private static final int SPINNER_VIEW_TYPE = 1;
    private static final int EPISODE_VIEW_TYPE = 2;
    private LazyLoadedEpisodeList lazyLoadedEpisodeList;
    private RowItemContract.EpisodePresenter presenter;
    private Context context;

    public PagedEpisodesAdapter(LazyLoadedEpisodeList lazyLoadedEpisodeList, Context context, EpisodeViewer episodeViewer) {
        this.lazyLoadedEpisodeList = lazyLoadedEpisodeList;
        presenter = new EpisodeListViewPresenter(episodeViewer);
        this.context = context;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup viewGroup, int viewType) {
        RecyclerView.ViewHolder result;
        if (viewType == EPISODE_VIEW_TYPE) {
            View view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.episode_listitem, viewGroup, false);
            result = new EpisodeViewHolder(view, context);
        } else {
            View view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.loading_listitem, viewGroup, false);
            result = new RecyclerView.ViewHolder(view) {
            };
        }
        return result;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof EpisodeViewHolder) {
            // set presenter
            EpisodeViewHolder viewHolder = (EpisodeViewHolder) holder;
            viewHolder.setPresenter(presenter);
            presenter.update(lazyLoadedEpisodeList.getEpisode(position), viewHolder);
        } else {
            lazyLoadedEpisodeList.loadEpisodes();
        }
    }

    @Override
    public int getItemCount() {
        int noOfItems = lazyLoadedEpisodeList.getNoOfLoadedItems();

        if (noOfItems == 0) {
            //just the loading spinner row
            noOfItems = 1;

        } else if (lazyLoadedEpisodeList.getNoOfLoadedItems() < lazyLoadedEpisodeList.getTotalNoOfItems()) {
            noOfItems++;
        }

        return noOfItems;
    }

    @Override
    public int getItemViewType(int positionInList) {
        int result;
        if (positionInList == getItemCount() - 1 && !lazyLoadedEpisodeList.allItemsLoaded()) {
            result = SPINNER_VIEW_TYPE;
        } else {
            result = EPISODE_VIEW_TYPE;
        }
        return result;
    }
}

File: app/src/main/java/se/asapehrsson/podtest/PagedEpisodesAdapter.kt
package se.asapehrsson.podtest

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup


class PagedEpisodesAdapter(private val lazyLoadedEpisodeList: LazyLoadedEpisodeList, private val context: Context, episodeViewer: EpisodeViewer) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    var presenter: EpisodeContract.Presenter = EpisodeListViewPresenter(episodeViewer)
    var firstTime: Boolean = true

    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val result: RecyclerView.ViewHolder
        if (viewType == EPISODE_VIEW_TYPE) {
            val view = LayoutInflater.from(viewGroup.context).inflate(R.layout.episode_listitem, viewGroup, false)
            result = EpisodeViewHolder(view, context)
        } else {
            val view = LayoutInflater.from(viewGroup.context).inflate(R.layout.loading_listitem, viewGroup, false)
            result = object : RecyclerView.ViewHolder(view) {
            }
        }
        return result
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is EpisodeViewHolder) {
            // set presenter
            val viewHolder = holder
            viewHolder.setPresenter(presenter)
            presenter.update(lazyLoadedEpisodeList.getEpisode(position), viewHolder)
        } else {
            lazyLoadedEpisodeList.loadEpisodes()
        }
    }

    override fun getItemCount(): Int {
        var noOfItems = lazyLoadedEpisodeList.noOfLoadedItems

        if (noOfItems == 0) {
            //just the loading spinner row
            noOfItems = 1

        } else if (lazyLoadedEpisodeList.noOfLoadedItems < lazyLoadedEpisodeList.totalNoOfItems) {
            noOfItems++
        }

        return noOfItems
    }

    override fun getItemViewType(positionInList: Int): Int {
        val result: Int
        if (positionInList == itemCount - 1 && !lazyLoadedEpisodeList.allItemsLoaded()) {
            result = SPINNER_VIEW_TYPE
        } else {
            result = EPISODE_VIEW_TYPE
        }
        return result
    }

    companion object {
        private val SPINNER_VIEW_TYPE = 1
        private val EPISODE_VIEW_TYPE = 2
    }
}
